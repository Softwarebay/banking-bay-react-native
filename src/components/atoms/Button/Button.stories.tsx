import { action } from '@storybook/addon-actions';
import { text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react-native';
import { ViewStory } from 'utils/storybook/ViewStory';

import { Button } from './Button';

storiesOf('Atoms/Button', module)
	.addDecorator((getStory) => <ViewStory>{getStory()}</ViewStory>)
	.add('Primary', () => (
		<Button
			testID="btn-test-id"
			title={text('title', 'Button')}
			type="primary"
			onPress={action('onPress')}
		/>
	))
	.add('Secondary', () => (
		<Button
			testID="btn-test-id"
			title={text('title', 'Button')}
			type="secondary"
			onPress={action('onPress')}
		/>
	))
	.add('Ternary', () => (
		<Button
			testID="btn-test-id"
			title={text('title', 'Button')}
			type="ternary"
			onPress={action('onPress')}
		/>
	));
