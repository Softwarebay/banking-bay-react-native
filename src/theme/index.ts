interface Colors {
	screenBackground: string;
	mainDark: string;
	bodyText: string;
	pink: string;
	buttonLight: string;
	buttonDark: string;
	green: string;
	white: string;
	shadow: string;
	grey600: string;
	grey300: string;
	grey100: string;
}

export const colors: Colors = {
	screenBackground: '#FAFCFF',
	mainDark: '#1B1D4D',
	bodyText: '#4C4C60',
	pink: '#FF5887',
	buttonLight: '#97DAFF',
	buttonDark: '#16579D',
	green: '#3EB290',
	white: '#FFFFFF',
	shadow: '#161966',
	grey600: '#868698',
	grey300: '#CED6E1',
	grey100: '#F1F5FD',
};

interface Shadows {
	ios: {
		shadowColor: string;
		shadowOffset: {
			width: number;
			height: number;
		};
		shadowOpacity: number;
		shadowRadius: number;
	};
	android: {
		elevation: number;
		shadowColor: string;
	};
}

export const shadows: Shadows = {
	ios: {
		shadowColor: colors.shadow,
		shadowOffset: {
			width: 0,
			height: 6,
		},
		shadowOpacity: 0.08,
		shadowRadius: 16,
	},
	android: {
		elevation: 4,
		shadowColor: colors.shadow,
	},
};
