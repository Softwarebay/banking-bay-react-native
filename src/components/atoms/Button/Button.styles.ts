import { StyleSheet } from 'react-native';
import { colors } from 'theme';

export const styles = StyleSheet.create({
	container: {
		alignItems: 'center',
		borderRadius: 10,
		height: 55,
		justifyContent: 'center',
	},
	pressed: {
		opacity: 0.8,
	},
	secondary: {
		backgroundColor: colors.grey100,
		borderColor: colors.grey300,
		borderWidth: 1,
	},
	secondaryText: {
		color: colors.mainDark,
	},
	ternary: {
		backgroundColor: colors.green,
		borderWidth: 0,
		height: 40,
	},
	ternaryText: {
		fontSize: 15,
	},
	text: {
		color: colors.white,
		fontFamily: 'MulishSemiBold',
		fontSize: 17,
	},
});
