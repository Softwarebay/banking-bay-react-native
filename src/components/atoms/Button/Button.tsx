import { Pressable, StyleProp, Text, ViewStyle } from 'react-native';
import { LinearGradient } from 'expo-linear-gradient';

import { colors } from 'theme';
import { styles } from './Button.styles';

export interface Props {
	title: string;
	type?: 'primary' | 'secondary';
	testID: string;
	onPress: () => void;
	style?: StyleProp<ViewStyle>;
}

export const Button = ({
	title,
	testID,
	style,
	type = 'primary',
	onPress,
}: Props) => {
	const secondaryContainerStyles =
		type === 'secondary' ? [styles.container, styles.secondary] : [];

	return (
		<Pressable
			style={({ pressed }) => [
				pressed && styles.pressed,
				...secondaryContainerStyles,
				style,
			]}
			testID={testID}
			onPress={onPress}
		>
			{type === 'secondary' && (
				<Text style={[styles.text, styles[`${type}Text`]]}>{title}</Text>
			)}

			{type === 'primary' && (
				<LinearGradient
					colors={[colors.buttonLight, colors.buttonDark]}
					end={{ x: 1, y: 1 }}
					start={{ x: 0, y: 0 }}
					style={styles.container}
				>
					<Text style={[styles.text, styles[`${type}Text`]]}>{title}</Text>
				</LinearGradient>
			)}
		</Pressable>
	);
};
