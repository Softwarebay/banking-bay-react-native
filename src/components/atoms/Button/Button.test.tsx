import { fireEvent, render, screen } from '@testing-library/react-native';
import testRenderer from 'react-test-renderer';

import { colors } from 'theme';
import { Button } from './Button';

import type { Props } from './Button';

describe('Button', () => {
	const mockOnPress = jest.fn<Props['onPress'], []>();
	const props: Props = {
		title: 'Button',
		onPress: mockOnPress,
		testID: 'btn-test-id',
	};
	const secondaryProps: Props = {
		...props,
		type: 'secondary',
	};

	it('should match snapshot', () => {
		const tree = testRenderer.create(<Button {...props} />).toJSON();

		expect(tree).toMatchSnapshot();
	});

	it('should display button text passed as prop', () => {
		render(<Button {...props} />);

		const el = screen.getByTestId('btn-test-id');

		expect(el).toHaveTextContent('Button');
	});

	it('should default to primary button', () => {
		render(<Button {...props} />);

		const el = screen.getByTestId('btn-test-id');

		expect(el).not.toHaveStyle({
			backgroundColor: colors.grey100,
			borderColor: colors.grey300,
		});
	});

	it('should display secondary button', () => {
		render(<Button {...secondaryProps} />);

		const el = screen.getByTestId('btn-test-id');

		expect(el).toHaveStyle({
			backgroundColor: colors.grey100,
			borderColor: colors.grey300,
		});
	});

	it('should handle onPress events', () => {
		render(<Button {...props} />);

		const el = screen.getByTestId('btn-test-id');

		fireEvent.press(el);

		expect(mockOnPress).toHaveBeenCalledTimes(1);
	});
});
