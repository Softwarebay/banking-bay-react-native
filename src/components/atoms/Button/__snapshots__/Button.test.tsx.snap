// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should match snapshot - primary button 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      false,
      undefined,
    ]
  }
  testID="btn-test-id"
>
  <ViewManagerAdapter_ExpoLinearGradient
    proxiedProperties={
      Object {
        "colors": Array [
          4288142079,
          4279654301,
        ],
        "endPoint": Array [
          1,
          1,
        ],
        "locations": undefined,
        "startPoint": Array [
          0,
          0,
        ],
      }
    }
    style={
      Object {
        "alignItems": "center",
        "borderRadius": 10,
        "height": 55,
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#FFFFFF",
          "fontFamily": "MulishSemiBold",
          "fontSize": 17,
        }
      }
    >
      Button
    </Text>
  </ViewManagerAdapter_ExpoLinearGradient>
</View>
`;

exports[`Button should match snapshot - secondary button 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      false,
      Object {
        "alignItems": "center",
        "borderRadius": 10,
        "height": 55,
        "justifyContent": "center",
      },
      Object {
        "backgroundColor": "#F1F5FD",
        "borderColor": "#CED6E1",
        "borderWidth": 1,
      },
      undefined,
    ]
  }
  testID="btn-test-id"
>
  <Text
    style={
      Array [
        Object {
          "color": "#FFFFFF",
          "fontFamily": "MulishSemiBold",
          "fontSize": 17,
        },
        Object {
          "color": "#1B1D4D",
        },
      ]
    }
  >
    Button
  </Text>
</View>
`;

exports[`Button should match snapshot - ternary button 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      false,
      Object {
        "alignItems": "center",
        "borderRadius": 10,
        "height": 55,
        "justifyContent": "center",
      },
      Object {
        "backgroundColor": "#3EB290",
        "borderWidth": 0,
        "height": 40,
      },
      undefined,
    ]
  }
  testID="btn-test-id"
>
  <Text
    style={
      Array [
        Object {
          "color": "#FFFFFF",
          "fontFamily": "MulishSemiBold",
          "fontSize": 17,
        },
        Object {
          "fontSize": 15,
        },
      ]
    }
  >
    Button
  </Text>
</View>
`;
